internal/controller/gardener_cluster_controller.go

```diff
@@ -162,16 +203,146 @@ func (r *GardenerClusterController) getSecret(shootName string) (*corev1.Secret,
 	return &secretList.Items[0], nil
 }
 
-func (r *GardenerClusterController) createSecret(ctx context.Context, cluster infrastructuremanagerv1.GardenerCluster) error {
-	secret, err := r.newSecret(cluster)
+func (controller *GardenerClusterController) createOrRotateKubeconfigSecret(ctx context.Context, cluster *imv1.GardenerCluster, lastSyncTime time.Time) (bool, error) {
+	existingSecret, err := controller.getSecret(cluster.Spec.Shoot.Name)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		cluster.UpdateConditionForErrorState(imv1.ConditionTypeKubeconfigManagement, imv1.ConditionReasonFailedToGetSecret, metav1.ConditionTrue, err)
+		return true, err
+	}
+
+	if !secretNeedsToBeRotated(cluster, existingSecret, controller.rotationPeriod) {
+		message := fmt.Sprintf("Secret %s in namespace %s does not need to be rotated yet.", cluster.Spec.Kubeconfig.Secret.Name, cluster.Spec.Kubeconfig.Secret.Namespace)
+		controller.log.Info(message, loggingContextFromCluster(cluster)...)
+		return false, nil
+	}
+
+	if secretRotationForced(cluster) {
+		message := fmt.Sprintf("Rotation of secret %s in namespace %s forced.", cluster.Spec.Kubeconfig.Secret.Name, cluster.Spec.Kubeconfig.Secret.Namespace)
+		controller.log.Info(message, loggingContextFromCluster(cluster)...)
+	}
+
+	kubeconfig, err := controller.KubeconfigProvider.Fetch(cluster.Spec.Shoot.Name)
+	if err != nil {
+		cluster.UpdateConditionForErrorState(imv1.ConditionTypeKubeconfigManagement, imv1.ConditionReasonFailedToGetKubeconfig, metav1.ConditionTrue, err)
+		return true, err
+	}
+
+	if existingSecret != nil {
+		return true, controller.updateExistingSecret(ctx, kubeconfig, cluster, existingSecret, lastSyncTime)
+	}
+
+	return true, controller.createNewSecret(ctx, kubeconfig, cluster, lastSyncTime)
+}
+
+func secretNeedsToBeRotated(cluster *imv1.GardenerCluster, secret *corev1.Secret, rotationPeriod time.Duration) bool {
+	return secretRotationTimePassed(secret, rotationPeriod) || secretRotationForced(cluster)
+}
+
+func secretRotationTimePassed(secret *corev1.Secret, rotationPeriod time.Duration) bool {
+	const rotationPeriodRatio = 0.95
+
+	if secret == nil {
+		return true
+	}
+
+	annotations := secret.GetAnnotations()
+
+	_, found := annotations[lastKubeconfigSyncAnnotation]
+
+	if !found {
+		return true
+	}
+
+	lastSyncTimeString := annotations[lastKubeconfigSyncAnnotation]
+	lastSyncTime, err := time.Parse(time.RFC3339, lastSyncTimeString)
+	if err != nil {
+		return true
+	}
+	now := time.Now()
+	alreadyValidFor := now.Sub(lastSyncTime)
+
+	return alreadyValidFor.Minutes() >= rotationPeriodRatio*rotationPeriod.Minutes()
```

Comment:
would it be possible to remove this rotationPeriodRatio parameter?

Reply:
without that, @akgalwas had a test case where instead after 36 minutes, rotation happened after ~72 minutes and if I recall correctly, for 12 minutes the kubeconfig was invalid.